User authentication with kerberos (#402)
stackabletech/hive-operator: docs/modules/hive/pages/usage-guide/security.adoc

= Security

== Authentication
Currently, the only supported authentication mechanism is Kerberos, which is disabled by default.
For Kerberos to work a Kerberos KDC is needed, which the users need to provide.
The xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation] states which kind of Kerberos servers are supported and how they can be configured.

=== 1. Prepare Kerberos server
To configure HDFS to use Kerberos you first need to collect information about your Kerberos server, e.g. hostname and port.
Additionally, you need a service-user which the secret-operator uses to create principals for the HDFS services.

=== 2. Create Kerberos SecretClass
The next step is to enter all the necessary information into a SecretClass, as described in xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation]. The following guide assumes you have named your SecretClass `kerberos`.

=== 3. Configure HDFS to use SecretClass
The next step is to configure your HdfsCluster to use the newly created SecretClass. Please follow the xref:hdfs:usage-guide/security.adoc[HDFS security guide] to set up and test this.

> Please make sure to use the SecretClass named `kerberos`.

Reviewer: Can we please also document the requirement of HMS impersonating the HMS clients?

So to be concrete the following settings:
https://github.com/stackabletech/hive-operator/blob/1f4ee1b4efb6147395258ba869b85166f2908b3d/tests/templates/kuttl/kerberos-hdfs/30-install-hdfs.yaml.j2#L28-L34

NOTE: This is only relevant if HDFS is used with the Hive metastore (many installations will use the metastore with an S3 backend instead of HDFS).

=== 4. Configure Hive to use SecretClass
The last step is to configure the same SecretClass for Hive, which is done similarly to HDFS.

IMPORTANT: HDFS and Hive need to use the same SecretClass (or at least use the same underlying Kerberos server).

[source,yaml]
----
spec:
  clusterConfig:
    authentication:

>       tlsSecretClass: tls # Optional, defaults to "tls"

Reviewer: ```suggestion
```

      kerberos:
        secretClass: kerberos # Put your SecretClass name in here
----

The `kerberos.secretClass` is used to give Hive the possibility to request keytabs from the secret-operator.

> The `tlsSecretClass` is needed to request TLS certificates, used e.g. for the Web UIs.

Reviewer: ```suggestion
```

= Security

== Authentication
Currently, the only supported authentication mechanism is Kerberos, which is disabled by default.
For Kerberos to work a Kerberos KDC is needed, which the users need to provide.
The xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation] states which kind of Kerberos servers are supported and how they can be configured.

=== 1. Prepare Kerberos server
To configure HDFS to use Kerberos you first need to collect information about your Kerberos server, e.g. hostname and port.
Additionally, you need a service-user which the secret-operator uses to create principals for the HDFS services.

=== 2. Create Kerberos SecretClass
The next step is to enter all the necessary information into a SecretClass, as described in xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation]. The following guide assumes you have named your SecretClass `kerberos`.

> === 3. Configure HDFS to use SecretClass

Author: d2ed5ab

The next step is to configure your HdfsCluster to use the newly created SecretClass. Please follow the xref:hdfs:usage-guide/security.adoc[HDFS security guide] to set up and test this.
Please make sure to use the SecretClass named `kerberos`.

=== 4. Configure Hive to use SecretClass
The last step is to configure the same SecretClass for Hive, which is done similarly to HDFS.

IMPORTANT: HDFS and Hive need to use the same SecretClass (or at least use the same underlying Kerberos server).

[source,yaml]
----
spec:
  clusterConfig:
    authentication:
      tlsSecretClass: tls # Optional, defaults to "tls"
      kerberos:
        secretClass: kerberos # Put your SecretClass name in here
----

The `kerberos.secretClass` is used to give Hive the possibility to request keytabs from the secret-operator.
The `tlsSecretClass` is needed to request TLS certificates, used e.g. for the Web UIs.

=== 5. Access Hive
In case you want to access Hive it is recommended to start up a client Pod that connects to Hive, rather than shelling into the master.

> We have an https://github.com/stackabletech/hive-operator/blob/main/tests/templates/kuttl/kerberos/70-install-access-hive.j2[integration test] for this exact purpose, where you can see how to connect and get a valid keytab.

Author: [18a0b7e](https://github.com/stackabletech/hive-operator/pull/402/commits/18a0b7eb115bcb75a93ef26454e7126b726c37ca)

> == Wire encryption

Author: d2ed5ab

For Kerberos to work a Kerberos KDC is needed, which the users need to provide.
The xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation] states which kind of Kerberos servers are supported and how they can be configured.

=== 1. Prepare Kerberos server
To configure HDFS to use Kerberos you first need to collect information about your Kerberos server, e.g. hostname and port.
Additionally, you need a service-user which the secret-operator uses to create principals for the HDFS services.

=== 2. Create Kerberos SecretClass
The next step is to enter all the necessary information into a SecretClass, as described in xref:home:secret-operator:secretclass.adoc#backend-kerberoskeytab[secret-operator documentation]. The following guide assumes you have named your SecretClass `kerberos`.

=== 3. Configure HDFS to use SecretClass
The next step is to configure your HdfsCluster to use the newly created SecretClass. Please follow the xref:hdfs:usage-guide/security.adoc[HDFS security guide] to set up and test this.
Please make sure to use the SecretClass named `kerberos`. It is also necessary to configure 2 additional things in HDFS:

> * define group mappings for user permissions with `hadoop.user.group.static.mapping.overrides`
> * tell HDFS that Hive will be "passing through" users: i.e. Hive does not need any _direct_ access permissions for itself, but should be able to impersonate Hive users when accessing HDFS. This is done by setting `hadoop.proxyuser.hive.users` and `hadoop.proxyuser.hive.hosts`.

Reviewer: ```suggestion
* Define group mappings for users with `hadoop.user.group.static.mapping.overrides`
* Tell HDFS that Hive is allowed to impersonate other users, i.e. Hive does not need any _direct_ access permissions for itself, but should be able to impersonate Hive users when accessing HDFS. This can be done by e.g. setting `hadoop.proxyuser.hive.users=*` and `hadoop.proxyuser.hive.hosts=*` to allow the user `hive`´to impersonate all other users.
```
